proposals/nnnn-inferring-senable-for-methods

```diff
@@ -0,0 +1,212 @@
+
+This should make `S` and its methods Sendable as well. However, when we pass our unapplied function `S.f`  to our generic function `g`, we get a warning that `S.f` is not Sendable as `g()` is expecting.  
+
+
+```
+struct S: P {
+  func f() { ... }
+}
+
+g(S.f) // Converting non-sendable function value to '@Sendable (S) -> (() -> Void)' may introduce data races
+```
+
+
+We can work around this by wrapping our unapplied function in a Sendable closure.  
+
+```
+// S.f($0) == S.f()
+g({ @Sendable **in** S.f($0) })
+```
+
+
+This is a lot of churn to get the expected behavior. The compiler should preserve `@Sendable` in the type signature instead.
+
+## Proposed solution
+
+For a function, the `@Sendable` attribute primarily influences the kinds of values that can be captured by the function. But methods of a nominal type do not capture anything but the object instance itself. Semantically, a method can be thought of as being represented by the following functions:
+
+
+```
+// Pseudo-code declaration of a Nominal Type:
+type NominalType {
+  func method(ArgType) -> ReturnType { /* body of method */ }
+}
+
+// Can desugar to these two global functions:
+func NominalType_method_partiallyAppliedTo(_ obj: NominalType) -> ((ArgType) -> ReturnType) {
+  let inner = { [obj] (_ arg1: ArgType) -> ReturnType in
+    return NominalType_method(obj, arg1)
+  }
+  return inner
+}
+func NominalType_method(_ self: NominalType, _ arg1: ArgType) -> ReturnType {
+  return self.method(arg1)
+}
+```
+
+Thus, the only way a partially-applied method can be `@Sendable` is if the `inner` closure were `@Sendable`, which is true if and only if the nominal type conforms to `Sendable`.
+
+
+```
+type NominalType : Sendable {
+  func method(ArgType) -> ReturnType { /* body of method */ }
+}
+```
+
+For example, by declaring the following type `Sendable`, the partial and unapplied function values of the type would have implied Sendabilty and the following code would compile with no errors. 
+
+```
+struct User : Sendable {
+  func updatePassword (new: String, old:String) -> Bool { /* update password*/ return true}
+}
+
+**let** unapplied: **@Sendable** (User) → ((String, String) → Bool) = User.updatePassword // no error
+
+**let** partial: **@Sendable** (String, String) → Bool = User().updatePassword // no error
+```
+
+
+
+## Detailed design
+
+This proposal includes five changes to `Sendable` behavior. 
+
+The first two are what we just discussed regarding partial and unapplied function values.
+
+```
+struct User : Sendable {
+  var address
+  var password
+  
+  func changeAddress () {/*do work*/ }
+}
+```
+
+1. The inference of `@Sendable` for unapplied references to methods of a Sendable type. 
+
+```
+**let** unapplied: **@Sendable** (User) → ((String, String) → Void) = User.changeAddress // no error
+```
+
+1. The inference of `@Sendable` for partially-applied methods of a Sendable type.
```

Comment:
Just checking: Is your implementation prepared to handle types which have a conditional conformance to Sendable? For instance, `someArray.sorted` should be `@Sendable` for an `Array<String>` but not for an `Array<NSObject>`.

Reply:
Yes the implementation handles this. I  can call this out in the proposal explicitly as well.